### PR TITLE
dialects: (llvm) Registering llvm.icmp

### PR DESCRIPTION
**Pull request by lfrenot**

### CODE REVIEW

File: tests/filecheck/dialects/llvm/arithmetic.mlir
Comment (tobiasgrosser) [inline tests/filecheck/dialects/llvm/arithmetic.mlir:75]: I guess you want examples for all predicate flags.

**Review by tobiasgrosser — APPROVED** [review]
Otherwise, this looks great to me. Maybe @superlopuh and @luisacicolini have comments?

**Comment by lfrenot** [comment]
I'm getting this error
```
# |     raise ParseError(at_position, msg)
# | xdsl.utils.exceptions.ParseError: stdin:26:72
# |   %icmp_eq = "llvm.icmp"(%arg0, %arg1) <{"predicate" = #llvm.predicateeq}> : (i32, i32) -> i32
# |                                                                         ^
# |                                                                         'llvm.predicateeq' is not registered
```
Since the predicate is supposed to be an i64 (according to `mlir-opt --mlir-print-op-generic`), I guess I have to change the way I'm representing it right now

File: xdsl/dialects/llvm.py
Comment (superlopuh) [inline xdsl/dialects/llvm.py:646]: There has to be a better way of doing this, I'd expect some helper to exist in the Python enum implementation, otherwise we should cache this mapping.

**Comment by superlopuh** [comment]
This PR adds an operation and an attribute, and only registers the new op in the dialect.

**Comment by lfrenot** [comment]
```
# | ValueError: 0 is not a valid ICmpPredicateFlag
```
I don't get why I'm getting this error with the current def of `ICmpPredicateFlag`

**Comment by lfrenot** [comment]
@superlopuh
Does it look good for you? 
Do you know how to properly do the enum thing?

File: xdsl/dialects/llvm.py
Comment (superlopuh) [inline xdsl/dialects/llvm.py:658]: why this change? it was better before as a strenum, no?

**Comment by superlopuh** [comment]
I would recommend reverting your last change, and looking into how to index the enum cases

**Comment by superlopuh** [comment]
```
from enum import StrEnum


class MyEnum(StrEnum):
    a = "a"
    b = "b"


MY_ENUM_CASES = list(MyEnum)

print(MY_ENUM_CASES[1])

```

This seems to work

**Comment by superlopuh** [comment]
and is not O(number of cases)

File: xdsl/dialects/llvm.py
Comment (superlopuh) [inline xdsl/dialects/llvm.py:633]:
```suggestion
    ALL_PREDICATES = tuple(ICmpPredicateFlag)

    @classmethod
    def from_int(cls, i: int):
        return ICmpPredicateAttr(cls.ALL_PREDICATES[i])
```

**Comment by lfrenot** [comment]
@superlopuh do you have any idea on how to fix the current error?

**Comment by superlopuh** [comment]
Ah no, it looks like I was too quick to approve. I would recommend marking the PR as draft if the tests don't pass, and reaching out on Zulip to iterate

**Comment by superlopuh** [comment]
I've actually never used this part of the API before, I'm happy to take a look in a bit if you're stuck, but it would be worth messaging on Zulip in case someone with more experience with your issue knows the answer directly. In this case pinging @alexarice would probably be a good idea, I think he's dealt with these attributes somewhat recently.

File: xdsl/dialects/llvm.py
Comment (alexarice) [inline xdsl/dialects/llvm.py:627]: I believe you want `OpaqueSyntaxAttribute` or `SpacedOpaqueSyntaxAttribute` here to get it to print properly

File: xdsl/dialects/llvm.py
Comment (alexarice) [inline xdsl/dialects/llvm.py:637]: Why is this marked as an abstract class?

File: xdsl/dialects/llvm.py
Comment (alexarice) [inline xdsl/dialects/llvm.py:682]: I think in general, print_string is preferred to print but this is a super nit

**Comment by lfrenot** [comment]
@alexarice Thank you for the suggestions, that worked

@superlopuh are we good to merge then?

**Review by alexarice — COMMENTED** [review]
I just tested the mlir version and it does this:
```
bin/mlir-opt ../../test.mlir --mlir-print-op-generic
"builtin.module"() ({
  %0:2 = "test.op"() : () -> (i32, i32)
  %1 = "llvm.icmp"(%0#0, %0#1) <{predicate = 0 : i64}> : (i32, i32) -> i1
}) : () -> ()
```
I feel the generic output of your function is not the same. Unsure if this is a blocker but I just wanted to point it out.

**Review by superlopuh — APPROVED** [review]
LGTM! The issue Alex pointed out is worth looking into, but I don't want to block on that necessarily. It would be great if you could take a quick look at the generic syntax before merging, and comment here.

**Comment by tobiasgrosser** [comment]
Let's merge this for now. @lfrenot, can you still check @alexarice's comment.

**Comment by tobiasgrosser** [comment]
> I just tested the mlir version and it does this:
> 
> ```
> bin/mlir-opt ../../test.mlir --mlir-print-op-generic
> "builtin.module"() ({
>   %0:2 = "test.op"() : () -> (i32, i32)
>   %1 = "llvm.icmp"(%0#0, %0#1) <{predicate = 0 : i64}> : (i32, i32) -> i1
> }) : () -> ()
> ```
> 
> I feel the generic output of your function is not the same. Unsure if this is a blocker but I just wanted to point it out.

@lfrenot, we should indeed be consistent with MLIR here. I just tried to use your icmp implementation to parse our test cases, but it failed as MLIR emits an integer attribute and you added your own attribute type. It would be great if this could be fixed.

**Pull request merged**
